### DISCUSSION TITLE
EEPROM library Vulnerability

### DISCUSSION THREAD

**Discussion by Btfriar6** [stm32duino/Arduino_Core_STM32#2485] [Libraries]
Dealing with an issue with if the MCU were to lose power while we are trying to update a field within the emulated EEPROM. We end up losing all the values in the emulated EEPROM. I understand that a prior to a write, you must erase the whole page. I'm wondering how much work it would be to use two flash pages for EEPROM emulation. One page would be used to hold the EEPROM, one would be used as a buffer. This would ensure that you would have a copy of your EEPROM even if you do lose power while updating a value. Is there a better way to ensure that your data is protected in this scenario. I see that in ST AN4894 they say that they use two flash memory pages. 
![image](https://github.com/user-attachments/assets/be8779b3-11e7-46ba-8b08-11e78eb45ab0)

Thanks!

**Comment by Btfriar6** [comment]
Just adding another problem we have just ran into. When using the hardware timer library with a periodic interrupt, you must insure that the interrupt is disabled while editing the EEPROM. We were having issues with the emulated EEPORM resetting itself due to this. Could this be added into the libraries functionality?

**Comment by fpistm** [comment]
Hi @Btfriar6 
Unfortunately, using another page could consume lot of flash depending of the series.
EEPROM is provided for Arduino compatibility and generic across all stm32 series. Honestly I don't advise to use it or  it should be used only to handle few data.
As the flash is used you should prevent to write several times due to flash lifetime and ensure no other operation are ongoing in parallel (irq, power,...).
